File: livecarddemo/src/main/java/com/gdkdemo/livecard/LiveCardDemoActivity.java
package com.gdkdemo.livecard;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.speech.RecognizerIntent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;

import com.gdkdemo.livecard.service.LiveCardDemoLocalService;
import com.gdkdemo.livecard.R;
import com.google.android.glass.touchpad.Gesture;
import com.google.android.glass.touchpad.GestureDetector;

import java.util.ArrayList;


// TBD:
// Add context menu
//
// ...


// The "main" activity...
public class LiveCardDemoActivity extends Activity
{
    // For tap events
    private GestureDetector mGestureDetector;

    // Service to handle liveCard publishing, etc...
    private boolean mIsBound = false;
    private LiveCardDemoLocalService glassGuardLocalService;
    private ServiceConnection serviceConnection = new ServiceConnection() {
        public void onServiceConnected(ComponentName className, IBinder service) {
            Log.d("onServiceConnected() called.");
            glassGuardLocalService = ((LiveCardDemoLocalService.LocalBinder)service).getService();
        }
        public void onServiceDisconnected(ComponentName className) {
            Log.d("onServiceDisconnected() called.");
            glassGuardLocalService = null;
        }
    };
    private void doBindService()
    {
        bindService(new Intent(this, LiveCardDemoLocalService.class), serviceConnection, Context.BIND_AUTO_CREATE);
        mIsBound = true;
    }
    private void doUnbindService() {
        if (mIsBound) {
            unbindService(serviceConnection);
            mIsBound = false;
        }
    }
    private void doStartService()
    {
        startService(new Intent(this, LiveCardDemoLocalService.class));
    }
    private void doStopService()
    {
        stopService(new Intent(this, LiveCardDemoLocalService.class));
    }


    @Override
    protected void onDestroy()
    {
        doUnbindService();
        super.onDestroy();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        Log.d("onCreate() called.");

        setContentView(R.layout.activity_livecarddemo);

        // For gesture handling.
        mGestureDetector = createGestureDetector(this);


        // TBD:
        // Create an asyncTask for publishing liveCard...
        // ????

        // Or,
        // Start LiveCardDemoIntentService and
        // delegate the liveCard publishing to LiveCardDemoIntentService.
        // ...

        // (Short-lived) IntentService does not work for us...
//        Intent ggIntent = new Intent(this, LiveCardDemoIntentService.class);
//        // trigger liveCard publishing ???   --> No, this is done in IntentService.onStart()...
//        ggIntent.putExtra(LiveCardDemoConstants.EXTRA_KEY_ACTION, LiveCardDemoConstants.GUARD_SERVICE_START);
//        this.startService(ggIntent);

        // We need a real service.
        // bind does not work. We need to call start() explicitly...
        // doBindService();
        doStartService();
        // TBD: We need to call doStopService() when user "closes" the app....
        // ...

    }


    @Override
    protected void onResume()
    {
        super.onResume();
        Log.d("onResume() called.");

        Bundle extras = getIntent().getExtras();
        ArrayList<String> voiceResults = null;
        if(extras != null) {
            voiceResults = extras.getStringArrayList(RecognizerIntent.EXTRA_RESULTS);
            // ...
        }

        // tbd...
        // ...

        // For live card menu handling
        openOptionsMenu();

    }


//    @Override
//    public int onStartCommand(Intent intent, int flags, int startId)
//    {
//        ArrayList<String> voiceResults = intent.getExtras()
//                .getStringArrayList(RecognizerIntent.EXTRA_RESULTS);
//        // ...
//        // TBD:
//        // ...
//
//
//        // ???
//        // return super.onStartCommand(intent, flags, startId);
//        // ...
//        return 0;
//    }


    // TBD:
    // Just use context menu instead of gesture ???
    // ...

    @Override
    public boolean onGenericMotionEvent(MotionEvent event)
    {
        if (mGestureDetector != null) {
            return mGestureDetector.onMotionEvent(event);
        }
        return false;
    }

    private GestureDetector createGestureDetector(Context context)
    {
        GestureDetector gestureDetector = new GestureDetector(context);
        //Create a base listener for generic gestures
        gestureDetector.setBaseListener( new GestureDetector.BaseListener() {
            @Override
            public boolean onGesture(Gesture gesture) {
                if(Log.D) Log.d("gesture = " + gesture);
                if (gesture == Gesture.TAP) {
                    // do something on tap
                    Log.i("Gesture.TAP");
                    handleGestureTap();
                    return true;
                } else if (gesture == Gesture.TWO_TAP) {
                    // do something on two finger tap
                    Log.i("Gesture.TWO_TAP");
                    handleGestureTwoTap();
                    return true;
                } else if (gesture == Gesture.SWIPE_RIGHT) {
                    // do something on right (forward) swipe
                    Log.i("Gesture.SWIPE_RIGHT");
                    handleGestureSwipeRight();
                    return true;
                } else if (gesture == Gesture.SWIPE_LEFT) {
                    // do something on left (backwards) swipe
                    Log.i("Gesture.SWIPE_LEFT");
                    handleGestureSwipeLeft();
                    return true;
                }
                return false;
            }
        });
//        gestureDetector.setFingerListener(new GestureDetector.FingerListener() {
//            @Override
//            public void onFingerCountChanged(int previousCount, int currentCount) {
//                // do something on finger count changes
//            }
//        });
//        gestureDetector.setScrollListener(new GestureDetector.ScrollListener() {
//            @Override
//            public boolean onScroll(float displacement, float delta, float velocity) {
//                // do something on scrolling
//
//                // ????
//                return false;
//            }
//        });
        return gestureDetector;
    }


    // TBD:

    private void handleGestureTap()
    {
        Log.d("handleGestureTap() called.");


    }

    private void handleGestureTwoTap()
    {
        Log.d("handleGestureTwoTap() called.");


    }

    private void handleGestureSwipeRight()
    {
        Log.d("handleGestureSwipeRight() called.");


    }

    private void handleGestureSwipeLeft()
    {
        Log.d("handleGestureSwipeLeft() called.");


    }



    // TBD:
    // Context menu
    // ...


    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        Log.d("onCreateOptionsMenu() called.");

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_livecarddemo_livecard, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        Log.d("onOptionsItemSelected() called.");

        // Handle item selection.
        switch (item.getItemId()) {
            case R.id.menu_stop_livecarddemo:
                doStopService();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onOptionsMenuClosed(Menu menu)
    {
        Log.d("onOptionsItemSelected() called.");

        // Nothing else to do, closing the activity.
        finish();
    }


}

File: xamarindemo/LiveCardDemo2/LiveCardDemoActivity.cs
using System;
using System.Linq;
using Android.App;
using Android.Content;
using Android.Glass.Timeline;
using System.Collections.Generic;
using Android.Glass.App;
using Android.Glass.Widget;
using Android.Glass.Touchpad;
using Android.OS;
using Android.Runtime;
using Android.Util;
using Android.Views;
using Android.Widget;

namespace LiveCardDemo2
{
	[Activity(Label = "@string/app_name", MainLauncher = true, Enabled = true)]
	[IntentFilter(new[] { "com.google.android.glass.action.VOICE_TRIGGER" })]
	[MetaData("com.google.android.glass.VoiceTrigger", Resource = "@xml/voiceinput_livecarddemo2")]
	public class LiveCardDemoActivity : Activity
	{
		// For tap event
		private Android.Glass.Touchpad.GestureDetector mGestureDetector;

		// Service to handle liveCard publishing, etc...
		private bool mIsBound = false;
		private static LiveCardDemoLocalService liveCardDemoLocalService;
		private static string _tag = "LiveCardDemo.LiveCardDemoActivity";

		private IServiceConnection serviceConnection = new MyServiceConnection();

		public class MyServiceConnection : IServiceConnection {
			#region IServiceConnection implementation
			public void OnServiceConnected(ComponentName className, IBinder service) {
				Log.Debug(_tag, "onServiceConnected() called.");
				liveCardDemoLocalService = ((LiveCardDemoLocalService.LocalBinder)service).GetService();
			}
			public void OnServiceDisconnected(ComponentName className) {
				Log.Debug(_tag, "onServiceDisconnected() called.");
				liveCardDemoLocalService = null;
			}
			#endregion
			#region IDisposable implementation
			public void Dispose ()
			{
				//throw new NotImplementedException ();
			}
			#endregion
			#region IJavaObject implementation
			public IntPtr Handle {
				get {
					return new IntPtr ();
				}
			}
			#endregion
		}

		private void DoBindService()
		{
			BindService(new Intent(this, typeof(LiveCardDemoLocalService)), serviceConnection, Bind.AutoCreate);
			mIsBound = true;
		}
		private void DoUnbindService() {
			if (mIsBound) {
				UnbindService(serviceConnection);
				mIsBound = false;
			}
		}
		private void DoStartService()
		{
			StartService(new Intent(this, typeof(LiveCardDemoLocalService)));
		}
		private void DoStopService()
		{
			StopService(new Intent(this, typeof(LiveCardDemoLocalService)));
		}

		protected override void OnDestroy()
		{
			DoUnbindService();
			// DoStopService();   // TBD: When do we call Stop service???
			base.OnDestroy();
		}

		protected override void OnCreate (Bundle bundle)
		{
			base.OnCreate (bundle);

			Log.Debug(_tag, "onCreate() called.");

			SetContentView(Resource.Layout.Activity_LiveCardDemo2);

			// For gesture handling.
			mGestureDetector = CreateGestureDetector (this);


			// Service to control the life cycle of a LiveCard.
			// bind() does not work. We need to call start() explicitly...
			// doBindService();
			DoStartService();
			// TBD: We need to call doStopService() when the user "closes" the app....
			// ...
		}

		protected override void OnResume()
		{
			base.OnResume();
			Log.Debug(_tag, "onResume() called.");
		}


		// TBD:
		// Just use context menu instead of gesture ???
		// ...

		public override bool OnGenericMotionEvent (MotionEvent e)
		{
			Log.Debug (_tag, e.Action.ToString());
			//e. == MotionEventActions. 
			if (mGestureDetector != null) {
				return mGestureDetector.OnMotionEvent (e);
			}
			return false;
		}

		private Android.Glass.Touchpad.GestureDetector CreateGestureDetector(Context context)
		{
			Android.Glass.Touchpad.GestureDetector gestureDetector = new Android.Glass.Touchpad.GestureDetector (context);
			//Create a base listener for generic gestures
			gestureDetector.SetBaseListener (new MyGestureDetector (this));

			return gestureDetector;
		}

		private class MyGestureDetector : Android.Glass.Touchpad.GestureDetector.IBaseListener
		{
			LiveCardDemoActivity _context;

			public MyGestureDetector (LiveCardDemoActivity context)
			{
				_context = context;
			}

			#region IBaseListener implementation
			public bool OnGesture (Gesture gesture)
			{
				Log.Debug (_tag, "gesture = " + gesture);
				if (gesture == Gesture.Tap) {
					_context.HandleGestureTap ();
					return true;
				} else if (gesture == Gesture.TwoTap) {
					_context.HandleGestureTwoTap ();
					return true;
				}
				return false;
			}
			#endregion
			#region IDisposable implementation
			public void Dispose ()
			{
			}
			#endregion
			#region IJavaObject implementation
			public IntPtr Handle {
				get {
					return new IntPtr ();
				}
			}
			#endregion
		}

		public void HandleGestureTap()
		{
			Log.Debug(_tag, "handleGestureTap() called.");
			DoStopService();
			Finish();
		}

		public void HandleGestureTwoTap()
		{
			Log.Debug(_tag, "handleGestureTwoTap() called.");
		}

//		public override bool OnCreateOptionsMenu (IMenu menu)
//		{
//			Log.Debug(_tag, "onCreateOptionsMenu() called.");
//
//			MenuInflater inflater = this.MenuInflater;
//			inflater.Inflate(Resource.Menu.menu_livecarddemo_livecard, menu);
//			return true;
//		}
//
//		public override bool OnOptionsItemSelected(IMenuItem item)
//		{
//			Log.Debug(_tag, "onOptionsItemSelected() called.");
//
//			// Handle item selection.
//			switch (item.ItemId) {
//			case Resource.Id.menu_stop_livecarddemo:
//				DoStopService();
//				return true;
//			default:
//				return base.OnOptionsItemSelected(item);
//			}
//		}
//		
//		public override void OnOptionsMenuClosed(IMenu menu)
//		{
//			Log.Debug(_tag, "onOptionsItemSelected() called.");
//
//			// Nothing else to do, closing the activity.
//			Finish();
//		}

//		private ServiceConnection serviceConnection = new ServiceConnection() {
//			public void OnServiceConnected(ComponentName className, IBinder service) {
//				Log.d("onServiceConnected() called.");
//				liveCardDemoLocalService = ((LiveCardDemoLocalService.LocalBinder)service).getService();
//			}
//			public void OnServiceDisconnected(ComponentName className) {
//				Log.d("onServiceDisconnected() called.");
//				liveCardDemoLocalService = null;
//			}
//		};




	}
}